fastestimator/trace/io/model_eval.py

```diff
@@ -0,0 +1,451 @@
+# Copyright 2019 The FastEstimator Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import inspect
+import json
+import locale
+import os
+import re
+import shutil
+from datetime import datetime
+from time import time
+from typing import Any, Callable, List, Union
+
+import numpy as np
+from pylatex import Command, Document, Itemize, MultiColumn, NoEscape, Package, Section, Subsection, Table, Tabularx, \
+    escape_latex
+
+import fastestimator as fe
+from fastestimator.trace.trace import Trace
+from fastestimator.util.data import Data
+from fastestimator.util.latex_util import TabularCell, WrapText
+from fastestimator.util.traceability_util import traceable
+from fastestimator.util.util import to_list, to_number
+
+
+@traceable()
+class TestCase():
+    """This class defines the test case that ModelEval trace will take to perform auto-testing.
+
+    Args:
+        description: A test description.
+        criteria: Function to perform the test returns True when the test passes and False when it fails. Input
+            variable names will be used as input keys to further derive input value.
+        aggregate: If True, this test is aggregate type and its criteria function will be examined at epoch_end. If
+            False, this test is per-instance type and its test criteria function will be examined at batch_end.
+        fail_threshold: Thershold of failure instance number to judge per-instance test as failed or passed. If failure
+            number is above this value, then the test fails; otherwise it passes. It only has effect when aggregate is
+            equal to False.
+    """
+    def __init__(self, description: str, criteria: Callable, aggregate: bool = True, fail_threshold: int = 0) -> None:
+        self.description = description
+        self.criteria = criteria
+        self.criteria_inputs = inspect.signature(criteria).parameters.keys()
+        self.aggregate = aggregate
+        if not self.aggregate:
+            self.fail_threshold = fail_threshold
+
+    def clean_result(self) -> None:
+        if not self.aggregate:
+            self.result = []
```

Comment:
all class member variables should appear within the init method

Reply:
to clarify, if someone is looking at only the init method, they should see all of the self.members there without going to any other functions